File: 3381-maximum-subarray-sum-with-length-divisible-by-k/3381-maximum-subarray-sum-with-length-divisible-by-k.cpp
#pragma GCC optimize("O3", "unroll-loops");
static const int __ = []() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    return 0;
}();
using ll = long long;
class Solution {
   public:
    long long maxSubarraySum(vector<int>& nums, int k) {
        int n = nums.size();
        vector<ll> minPref(k, LLONG_MAX);
        ll ans = LLONG_MIN;
        ll pref = 0;
        for (int i = 0; i < n; i++) {
            pref += nums[i];
            int r = (i + 1) % k;

            if ((i + 1) % k == 0) ans = max(ans, pref);
            if (minPref[r] != LLONG_MAX) {
                ans = max(ans, pref - minPref[r]);
            }

            minPref[r] = min(minPref[r], pref);
        }

        return ans;
    }
};

File: 3381-maximum-subarray-sum-with-length-divisible-by-k/3381-maximum-subarray-sum-with-length-divisible-by-k.rs
impl Solution {
    pub fn max_subarray_sum(nums: Vec<i32>, k: i32) -> i64 {
        let k = k as usize;
        let mut ans = i64::MIN;
        let mut minPref = vec![i64::MAX; k];
        let mut pref = 0_i64;

        for i in 0..nums.len() {
            pref += nums[i] as i64;
            let r = (i + 1) % k;

            if (i + 1) % k == 0 {
                ans = ans.max(pref);
            }
            if minPref[r] != i64::MAX {
                ans = ans.max(pref - minPref[r]);
            }
            minPref[r] = minPref[r].min(pref);
        }

        ans
    }
}